text/0000-guesses.md

```diff
@@ -0,0 +1,258 @@
+remove the ampersand.
+
+When there is no clear solution (e.g. when the user references a type `Iter`,
+but there is no such type in the scope) heuristics can supply "guesses" (e.g.
+right now, the compiler supplies an unstructured list of types called `Iter`).
+
+# Detailed design
+[design]: #detailed-design
+
+## Diagnostics frontend API changes
+
+The compiler diagnostics API is extended with the following methods:
+
+```rust
+pub fn span_guesses<I>(&mut self, sp: Span, msg: &str, guesses: I) -> &mut Self where I: IntoIterator<Item = String>;
+pub fn span_guess(&mut self, sp: Span, msg: &str, guess: String) -> &mut Self {
+    self.span_guesses(sp, msg, ::std::iter::once(guess))
+}
+```
+
+`span_guess` is just a convenience method which forwards to `span_guesses`.
+
+`span_guesses` takes an iterator of guesses to be presented to the user.
+The span `sp` is the region of code which will be replaced by one of the code snippets
+given by the `guesses` iterator. The span can be zero bytes long, if the text is supposed
+to be inserted (e.g. adding a new `use xyz::abc;` at the top of the file) instead of
+replacing existing code.
+
+If multispan replacements (replacing multiple separate spans at once, e.g. when modifying the
+pattern in a for loop together with the object being iterated over) are desired,
+one should modify the `Diagnostic` object manually. This might be inconvenient, but
+multispan replacements are very rare (occurring exactly
+[twice in clippy](https://github.com/Manishearth/rust-clippy/search?utf8=%E2%9C%93&q=multispan_sugg))
+and therefore the API should be figured out once/if they become more common.
+
+## Changes in diagnostic API usage
+
+### Replace `suggestions` with `guesses`
+
+All current calls to `span_suggestion` are replaced with `span_guess` if 
+their replacement is obtained by a heuristic. A new type of tests is added: "suggestion". Tests in the "suggestion" directory must be failing
+with an error (or denied lint) and contain a suggestion (not required to
+be emitted by the error or lint failing the test). Then the suggestion is
+applied, and the test must now compile and run successfully.
+
+### Change `help` messages containing code snippets to `guesses`
+
+Whenever a `span_help` is passed a snippet or hand built expression,
+it is changed into a `span_guess` or `span_guesses` if multiple helps
+are generated in some form of a loop. If `span_guesses` is used, all
+arbitrary limits on the number of displayed items is removed. This limit
+is instead enforced by the command line diagnostic backend.
+
+## Json diagnostics backend API changes
+
+The json object gets a new field `guesses` which is an array of
+`Guess` objects. Every `Guess` object contains an array of span +
+snippet pairs stating which part of the code should be replaced
+with what replacment. There is no need for all guesses to replace
+the same piece of code or even require the same number of replacements.
```

Comment:
It would be better for rustfix to directly use the Diagnostics API like [RLS wants to do](https://github.com/rust-lang-nursery/rls/issues/25)